packages/user-operation-controller/src/utils/gas.ts

```diff
@@ -0,0 +1,86 @@
+import { hexToBN } from '@metamask/controller-utils';
+import { BN, addHexPrefix } from 'ethereumjs-util';
+
+import { ENTRYPOINT } from '../constants';
+import { Bundler } from '../helpers/Bundler';
+import { createModuleLogger, projectLogger } from '../logger';
+import type {
+  PrepareUserOperationResponse,
+  UserOperationMetadata,
+} from '../types';
+
+const log = createModuleLogger(projectLogger, 'gas');
+
+/**
+ * A multiplier to apply to all gas estimate values returned from the bundler.
+ */
+const GAS_ESTIMATE_MULTIPLIER = 1.5;
+
+/**
+ * Populates the gas properties for a user operation.
+ * @param metadata - The metadata for the user operation.
+ * @param prepareResponse - The prepare response from the smart contract account.
+ */
+export async function updateGas(
+  metadata: UserOperationMetadata,
+  prepareResponse: PrepareUserOperationResponse,
+) {
+  const { userOperation } = metadata;
+
+  if (prepareResponse.gas) {
+    userOperation.callGasLimit = prepareResponse.gas.callGasLimit;
+    userOperation.preVerificationGas = prepareResponse.gas.preVerificationGas;
+    userOperation.verificationGasLimit =
+      prepareResponse.gas.verificationGasLimit;
+
+    log('Using gas values from smart contract account', {
+      callGasLimit: userOperation.callGasLimit,
+      preVerificationGas: userOperation.preVerificationGas,
+      verificationGasLimit: userOperation.verificationGasLimit,
+    });
+
+    return;
+  }
+
+  const payload = {
+    ...userOperation,
+    callGasLimit: '0x1',
+    preVerificationGas: '0x1',
+    verificationGasLimit: '0x1',
+  };
+
+  const bundler = new Bundler(metadata.bundlerUrl as string);
+  const estimate = await bundler.estimateUserOperationGas(payload, ENTRYPOINT);
+
+  userOperation.callGasLimit = normalizeGasEstimate(estimate.callGasLimit);
+  userOperation.preVerificationGas = normalizeGasEstimate(
+    estimate.preVerificationGas,
+  );
+  userOperation.verificationGasLimit = normalizeGasEstimate(
+    (estimate.verificationGasLimit ?? estimate.verificationGas) as
+      | string
+      | number,
```

Comment:
How do we know we can make this cast (ie, remove the undefined)? 

We could also do the following just to be sure:
estimate.verificationGasLimit ?? estimate.verificationGas ?? 0

Reply:
We're assuming the bundler response is valid for now and the spec includes the `verificationGasLimit` property in the response.

We only have the fallback to use the `verificationGas` property since I noticed some public bundlers were using that property instead.

We can add futher validation in future PRs, but for now we're assuming the provided bundler responses are valid.